### PR TITLE
Add Dactyl Manuform 6x6_4 and Kinesis keyboards

### PR DESCRIPTION
**Pull request by dmik**
## Description

This pull request adds two new keyboards to `handwired/dactyl_manuform`: `6x6_4` and `6x6_kinesis`.  The first one is based on `6x6` but it adds 4 new keys to the bottom alpha-numeric row increasing the total number of keys from 76 to 80. The second one is based on `6x6_4` but it assumes the physical key layout like on the Kinesis Advantage keyboard (and it is the primary reason `6x6_4` was hacked out — Kinesis Advantage needs a bit more keys).

It's worth to mention that the functional row of the original Kinesis Advantage keyboard is very specific and it is not fully implemented by this keyboard. The function keys occupy the top row but there is room only for 10 F-keys (because of ESC living to the left of F1) — F11 and F12 go to the second layer. 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

### CODE REVIEW

File: keyboards/handwired/dactyl_manuform/6x6_4/keymaps/default/config.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/*
```
Comment (drashna) [inline keyboards/handwired/dactyl_manuform/6x6_4/keymaps/default/config.h:1]: This entire file can be removed, as the defines are ... the default behavior.

File: keyboards/handwired/dactyl_manuform/6x6_4/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,46 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/handwired/dactyl_manuform/6x6_4/keymaps/default/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

File: keyboards/handwired/dactyl_manuform/6x6_4/6x6_4.c [hunk 3]
```diff
@@ -0,0 +1 @@
+#include "6x6_4.h"
```
Comment (drashna) [inline keyboards/handwired/dactyl_manuform/6x6_4/6x6_4.c:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

File: keyboards/handwired/dactyl_manuform/6x6_4/6x6_4.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
+#pragma once
```
Comment (drashna) [inline keyboards/handwired/dactyl_manuform/6x6_4/6x6_4.h:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

File: keyboards/handwired/dactyl_manuform/6x6_4/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+#define _QWERTY 0
+#define _LOWER 1
+#define _RAISE 2
```
Comment (drashna) [inline keyboards/handwired/dactyl_manuform/6x6_4/keymaps/default/keymap.c:5]:
```suggestion
enum custom_layers {
    _QWERTY,
    _LOWER,
    _RAISE,
);
```

File: keyboards/handwired/dactyl_manuform/6x6_kinesis/6x6_kinesis.c [hunk 6]
```diff
@@ -0,0 +1 @@
+#include "6x6_kinesis.h"
```
Comment (drashna) [inline keyboards/handwired/dactyl_manuform/6x6_kinesis/6x6_kinesis.c:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

File: keyboards/handwired/dactyl_manuform/6x6_kinesis/6x6_kinesis.h [hunk 7]
```diff
@@ -0,0 +1,33 @@
+#pragma once
```
Comment (drashna) [inline keyboards/handwired/dactyl_manuform/6x6_kinesis/6x6_kinesis.h:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

File: keyboards/handwired/dactyl_manuform/6x6_kinesis/keymaps/default/config.h [hunk 8]
```diff
@@ -0,0 +1,25 @@
+/*
```
Comment (drashna) [inline keyboards/handwired/dactyl_manuform/6x6_kinesis/keymaps/default/config.h:1]: File should be removed.

File: keyboards/handwired/dactyl_manuform/6x6_kinesis/keymaps/default/keymap.c [hunk 9]
```diff
@@ -0,0 +1,46 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/handwired/dactyl_manuform/6x6_kinesis/keymaps/default/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

File: keyboards/handwired/dactyl_manuform/6x6_kinesis/keymaps/default/keymap.c [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+#define _QWERTY 0
+#define _LOWER 1
+#define _RAISE 2
```
Comment (drashna) [inline keyboards/handwired/dactyl_manuform/6x6_kinesis/keymaps/default/keymap.c:5]:
```suggestion
enum custom_layers {
    _QWERTY,
    _LOWER,
    _RAISE,
);
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by dmik** [comment]
Thanks for the feedback, I guess I fixed it all.

BTW is there a way to test info.json other than publish it at https://config.qmk.fm? I'm worried if I guessed it all right.

**Comment by drashna** [comment]
If you hit Ctrl + Shft + I on the Configurator, it will let you load an info.json.

**Comment by dmik** [comment]
This doesn't appear to work in Safari under macOS. Is there any other way?

**Comment by drashna** [comment]
It should, but you have to hit control, then shift, then I, in that order, and one at a time.  It's ... very picky.  And I can confirm that it works in Safari under MacOS>  Though, safari isn't officially supported, from what I recall.

**Comment by dmik** [comment]
Great, it worked, thanks. Fixed one minor glitch.

Okay, do I also need to provide `keymap.json` to show the default layout in the UI or it is somehow generated from a respective `keymap.c` automatically?

Re compatibility, yeah there is a banner about using an unsupported browser when under Safari but everything appears to work very well (Safari 15.1 at least).

**Comment by drashna** [comment]
Default layout for QMK Configurator is added to that repo, and is done afterwards

https://docs.qmk.fm/#/configurator_default_keymaps

**Comment by dmik** [comment]
Okay, I see. So do I need to do anything else within this pull request?

**Comment by dmik** [comment]
Could anyone please tell me what else needs to be done for this pull request to be merged in and appear on the configurator page? @drashna?

File: keyboards/handwired/dactyl_manuform/6x6_4/rules.mk [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = no       # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
```
Comment (drashna) [inline keyboards/handwired/dactyl_manuform/6x6_4/rules.mk:15]:
```suggestion
```

File: keyboards/handwired/dactyl_manuform/6x6_kinesis/rules.mk [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = no       # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
```
Comment (drashna) [inline keyboards/handwired/dactyl_manuform/6x6_kinesis/rules.mk:15]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by drashna** [comment]
Also, would prefer to see bootmagic enabled, as it can be a pita to get to the bootloader  boards like these.

**Comment by dmik** [comment]
@drashna Committed your suggestions.

> Also, would prefer to see bootmagic enabled, as it can be a pita to get to the bootloader boards like these.

Well, I don't have such an experience as all boards I've dealt with have physical reset pins. So unless you insist, I would leave it as is (those who need it will enable it on their own). Also considering that bootmgic is disabled for all other Dactyl Manuform variants out there.

**Review by drashna — APPROVED** [review]

File: keyboards/handwired/dactyl_manuform/6x6_kinesis/rules.mk [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = no       # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+NKRO_ENABLE = no            # USB Nkey Rollover
```
Comment (fauxpark) [inline keyboards/handwired/dactyl_manuform/6x6_kinesis/rules.mk:15]:
```suggestion
NKRO_ENABLE = no            # Enable N-Key Rollover
```

File: keyboards/handwired/dactyl_manuform/6x6_4/rules.mk [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = no       # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+NKRO_ENABLE = no            # USB Nkey Rollover
```
Comment (fauxpark) [inline keyboards/handwired/dactyl_manuform/6x6_4/rules.mk:15]:
```suggestion
NKRO_ENABLE = no            # Enable N-Key Rollover
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Comment by drashna** [comment]
Thanks!

**Comment by DuMaM** [comment]
@dmik  Why don't you just added new keymap and new layer to 6x6?

**Comment by DuMaM** [comment]
That's not problem. My also is different one, take a look on 6x6.h and keymaps
There is a 5thumb one where there are missing 2 keys from thumb cluster and it works flawlessly.

**Comment by DuMaM** [comment]
@dmik If I make PR which adds your keyboard to Dactyl Manuform 6x6 folder, would you test it for me?

**Comment by dmik** [comment]
@DuMaM Ok now I see what you mean. But this example was not there when I did mine so I had no idea it would be picked up by the UI this way (I modelled mine after the 5x6 series). My 6x6_4 would definitely fit as a separate layout into 6x6 series now but I'm not sure about 6x6_kinesis because it also has a different geometry (physical key layout). What happens if `info.json` will contain both `LAYOUT_6x6` and, say, `LAYOUT_6x6_kinesis`? Will the UI select the correct one when selecting `LAYOUT_6x6_kinesis` merged into 6x6 according to your scheme?

If the answer is yes — then sure, go on with your PR, I will review and test it

P.S. I wonder why keymaps are completely ignored by the UI. How to have it recognise, say, `keymaps/default/keymap.c`? Is it possible to do automatically or one needs to create `keymap.json` with a copy of `keymap.c` data as done for e.g. `5x6_2_5/keymaps/default`? From that file it looks like it can be generated from `keymap.c` in an unattended way...

**Comment by DuMaM** [comment]
> I'm not sure about 6x6_kinesis because it also has a different geometry (physical key layout). What happens if info.json will contain both LAYOUT_6x6 and, say, LAYOUT_6x6_kinesis? Will the UI select the correct one when selecting LAYOUT_6x6_kinesis merged into 6x6 according to your scheme?
I guess by UI you mean QMK Configurator. It will be loaded without issues. The only problem is qmk configurator will load default keymap instead of yours. :disappointed:. 
If you referring to [toolbox](https://github.com/qmk/qmk_toolbox/releases) the answer is yes. 

> P.S. I wonder why keymaps are completely ignored by the UI (I see N/A for all keys when I select one of my layouts). How to make it recognise, say, keymaps/default/keymap.c? Is it possible to do automatically or one needs to create keymap.json with a copy of keymap.c data as done for e.g. 5x6_2_5/keymaps/default? From that file it looks like it can be generated from keymap.c in an unattended way...
https://github.com/qmk/qmk_firmware/pull/15891 Here is fix for this N/A problems. I recently added blackpill support and it was broken by this change.

**Comment by DuMaM** [comment]
@dmik It looks it could be looked up in ui. So when somebody will pick your layout, your keymap will be loaded :)

**Comment by dmik** [comment]
@DuMaM it doesn't seem to work this way. When I select my layout, the default keymap is not loaded... May be I didn't do something which is necessary for that, I don't know what.

**Comment by DuMaM** [comment]
https://github.com/qmk/qmk_firmware/pull/15891#pullrequestreview-853580647 
Take a look on this comment.

**Comment by dmik** [comment]
@DuMaM ok I see. And btw selecting the default keymap is more than fine. By `mine` I meant my keyboards (6x6_4 and 6x6_kinesis). If `keymaps/default` will be loaded for them - it will be exactly what I'd expect to.

**Comment by DuMaM** [comment]
@dmik yep, I will now try to fix this, please give me some time to do this and then we try to merge it into 6x6 :wink:

**Comment by dmik** [comment]
@DuMaM sure take your time! :) I'm ready to test.

**Pull request merged**